### ISSUE TITLE
SUM formula only calculating first cell in provided range

### ISSUE THREAD

**Issue by dvelopin** [PHPOffice/PhpSpreadsheet#1230]
This is:

```
- [X] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?

Cells with formula using SUM should return a subtotal of the cells from the provided range.

### What is the current behavior?

The returned value is only the value of the first cell from the provided range.

### What are the steps to reproduce?

Cell A11 value example: =SUM(A1:A10). Only value of A1 is returned.

Please provide a [Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

```php
<?php

require __DIR__ . '/vendor/autoload.php';

// Create new Spreadsheet object
$spreadsheet = new \PhpOffice\PhpSpreadsheet\Spreadsheet();

// add code that show the issue here...

$reader = IOFactory::createReader('Xls);
$reader->setLoadSheetsOnly('Sheet Name');
$spreadsheet = $reader->load('my-file.xls);
$sheetData = $spreadsheet->getActiveSheet()->toArray(null, true, true, true);

foreach($sheetData as $rowNumber => $rowData) {

	// For each row
	$valueINeed = $rowData['K']; // Column 11
	print $valueINeed;
}
```

### Which versions of PhpSpreadsheet and PHP are affected?

Latest version (1.9.0)

**Comment by dvelopin** [comment]
Looks like performing my example specifically it works fine. The same principle of SUM(A1:A10) in a much larger spreadsheet and more columns only returns the first column value, doing more testing.

**Comment by dvelopin** [comment]
They do, and thats the point of this report. Some rows are like:

0 0 35 5 <-- So the sum should be 40 but is 0
35 5 0 0 <-- So the sum should be 40 but is 35

This isn't happening with all rows, though. Once I saved this as a XLSX it worked just fine. Within excel the formula is working just fine regardless of XLS or XLSX. So I'm not sure there the trouble is in all honesty. For kicks I even tried changing field types to see if it had any affect but it didn't. I have been able to get passed my issue, but am not sure why with the XLS version it still wasn't working for some rows.

**Comment by dvelopin** [comment]
Yeah great point. The values are all from an external file and being parsed by PHP and then converted to an array to be worked with. No data manipulation is going on in the worksheet after I've converted it to an array.

To perform a quick test I did a new PHP script to simply load the file, convert to an array, and print that array out to the terminal. At this point the SUM works for most rows but not all of them. Since I appear to be the only one having this issue it may be something in my file. I think you're probably safe to just close this out. But it's up to you.

**Comment by MarkBaker** [comment]
I'll leave it open for a while, in case you find anything further, or anybody else experiences the issue.... 

note that there is also a formula debugger built into PHPSpreadsheet itself that may help.

I typically use:
```function evaluate(Spreadsheet $spreadSheet, Worksheet $workSheet, string $cell)
{
// Initialise the calculation engine for debug logging
    $calculationEngine = Calculation::getInstance($spreadSheet);
    $debugLog = $calculationEngine->getDebugLog();

    $calculationEngine->flushInstance();
    $debugLog->setWriteDebugLog(true);

    $formulaValue = $workSheet->getCell($cell)->getValue();

    echo PHP_EOL, 'Formula value for evaluation is ', $formulaValue, PHP_EOL;

    $canExecuteCalculation = false;
    try {
        $tokens = $calculationEngine->parseFormula($formulaValue, $workSheet->getCell($cell));
        echo 'Parser Stack :-', PHP_EOL;
        print_r($tokens);
        $canExecuteCalculation = true;
    } catch (Exception $e) {
        echo "PARSER ERROR: ", $e->getMessage(), PHP_EOL;

        echo 'Parser Stack :-';
        print_r($tokens);
    }

    $callStartTime = microtime(true);

    if ($canExecuteCalculation) {
        //  calculate
        try {
            $cellValue = $workSheet->getCell($cell)->getCalculatedValue();

            echo 'Result is ', $cellValue, PHP_EOL;

            echo 'Evaluation Log:', PHP_EOL;
            print_r($debugLog->getLog());
        } catch (Exception $e) {
            echo 'CALCULATION ENGINE ERROR: ', $e->getMessage(), PHP_EOL;

            echo 'Evaluation Log:', PHP_EOL;
            print_r($debugLog->getLog());
        }
    }

    $callEndTime = microtime(true);
    $callTime = $callEndTime - $callStartTime;

    echo PHP_EOL;
    echo 'Call time to evaluate formula was ', sprintf('%.4f', $callTime), ' seconds', PHP_EOL;
}
```
to help debug formula calculations

**Comment by dvelopin** [comment]
Thank you Mark, I'll definitely try that out. I appreciate it.

**Comment by dvelopin** [comment]
Mark,

Not sure why this is happening, but according to the debug code the formula value for evaluation is:

```
=SUM(T699:T699)
```

This is incorrect. If I open the spreadsheet in excel the value is:

```
=SUM(T699:X699)
```

Not sure why the formula is changing when being parsed.

```
Formula value for evaluation is =SUM(T699:T699)
Parser Stack :-
Array
(
    [0] => Array
        (
            [type] => Cell Reference
            [value] => T699
            [reference] => T699
        )

    [1] => Array
        (
            [type] => Cell Reference
            [value] => T699
            [reference] => T699
        )

    [2] => Array
        (
            [type] => Binary Operator
            [value] => :
            [reference] =>
        )

    [3] => Array
        (
            [type] => Operand Count for Function SUM()
            [value] => 1
            [reference] =>
        )

    [4] => Array
        (
            [type] => Function
            [value] => SUM(
            [reference] =>
        )

)
Result is 625
Evaluation Log:
Array
(
    [0] => Testing cache value for cell CustomerLoadList - Merge Repor!S699
    [1] => CustomerLoadList - Merge Repor!S699 => Evaluating Cell T699 in current worksheet
    [2] => CustomerLoadList - Merge Repor!S699 => Evaluation Result for cell CustomerLoadList - Merge Repor!T699 is a floating point number with a value of 625
    [3] => CustomerLoadList - Merge Repor!S699 => Evaluating Cell T699 in current worksheet
    [4] => CustomerLoadList - Merge Repor!S699 => Evaluation Result for cell CustomerLoadList - Merge Repor!T699 is a floating point number with a value of 625
    [5] => CustomerLoadList - Merge Repor!S699 => Evaluating Range "CustomerLoadList - Merge Repor!T699" : "CustomerLoadList - Merge Repor!T699"
    [6] => CustomerLoadList - Merge Repor!S699 => Evaluating Function SUM() with 1 argument
    [7] => CustomerLoadList - Merge Repor!S699 => Evaluating SUM( 625 )
    [8] => CustomerLoadList - Merge Repor!S699 => Evaluation Result for SUM() function call is a floating point number with a value of 625
)

Call time to evaluate formula was 0.0002 seconds
```

Looking at this it now makes sense why the value is coming back as the first cell in the range, but why is the formula changing for some, not all, cells?

**Comment by MarkBaker** [comment]
It isn't changing when being parsed:
`Formula value for evaluation is =SUM(T699:T699)` is the formula read from the cell __before__ being parsed; so it's the value being loaded from the file

**Comment by dvelopin** [comment]
Right. If I open the excel file its different than what it is seeing when being loaded from the file in PHP. Make sense? If I open the file, making absolutely no changes, and go to that cell, the formula is different. Saving the file as a xlsx loads just fine, though. So thats what I'm trying to figure out, is why?

**Comment by Pillem** [comment]
I have the same issue. I have a `.xls` file with in column `E` the formula `=SUM(B11:D11)`.  It gives me a result that only consists of the first value in the range. So the value of `B11`.

Using the debugging script gives me the same result that @dvelopin had where the parameters of the function are changed: `=SUM(B11:B11)`

I have also tried using `->getCalculatedValue()` which yields the same false result. Using `->getOldCalculatedValue()` does give me the correct result but this means not being able to get all calculated values with `->toArray(null, true, false, true)` and according to the documentation this could be risky:

> Note that this value is not guaranteed to reflect the actual calculated value because it is possible that auto-calculation was disabled in the original spreadsheet, and underlying data values used by the formula have changed since it was last calculated.

Saving the file as `.xlsx` makes it work as expected which is a relatively easy workaround.

**Comment by oleibman** [comment]
This is almost certainly solved by PR #4140 which was merged yesterday.
